e2e/rgraph_update_action_test.go

```diff
@@ -0,0 +1,347 @@
+	}
+	checkGCEResources(t, ctx, hcID, bsID, 120, 100)
+}
+func TestUpdateBackendService(t *testing.T) {
+	ctx := context.Background()
+	graphBuilder := rgraph.NewBuilder()
+	hcName := "hc-update"
+	hcID, err := buildHealthCheck(graphBuilder, hcName, 15)
+	if err != nil {
+		t.Fatalf("buildHealthCheck(_, %s, _) = %v, want nil", hcName, err)
+	}
+	bsName := "bs-update"
+	bsID, err := buildBackendService(graphBuilder, bsName, hcID, 80)
+	if err != nil {
+		t.Fatalf("buildBackendService(_, %s, _, 80) = %v, want nil", bsName, err)
+	}
+
+	graph, err := graphBuilder.Build()
+	if err != nil {
+		t.Fatalf("graphBuilder.Build() = %v, want nil", err)
+	}
+	result, err := plan.Do(ctx, theCloud, graph)
+	if err != nil {
+		t.Fatalf("plan.Do(_, _, _) = %v, want nil", err)
+	}
+
+	var viz exec.GraphvizTracer
+	ex, err := exec.NewSerialExecutor(result.Actions, exec.TracerOption(&viz))
+	if err != nil {
+		t.Logf("exec.NewSerialExecutor(_, _) err: %v", err)
+		return
+	}
+	res, err := ex.Run(context.Background(), theCloud)
+	if err != nil || res == nil {
+		t.Errorf("ex.Run(_,_) = %v, want nil", err)
+	}
+
+	t.Cleanup(func() {
+		err = theCloud.BackendServices().Delete(ctx, bsID.Key)
+		if err != nil {
+			t.Logf("delete backend service: %v", err)
+		}
+		err := theCloud.HealthChecks().Delete(ctx, hcID.Key)
+		if err != nil {
+			t.Logf("delete health check: %v", err)
+		}
+	})
+	checkGCEResources(t, ctx, hcID, bsID, 15, 80)
+
+	// update BackendService
+	bsID, err = buildBackendService(graphBuilder, bsName, hcID, 100)
+	if err != nil {
+		t.Fatalf("buildBackendService(_, %s, _, 100) = (_, %v), want (_, nil)", bsName, err)
+	}
+	graph, err = graphBuilder.Build()
+	if err != nil {
+		t.Fatalf("graphBuilder.Build() = %v, want nil", err)
+	}
+	result, err = plan.Do(ctx, theCloud, graph)
+	if err != nil {
+		t.Fatalf("plan.Do(_, _, _) = %v, want nil", err)
+	}
+
+	expectedActions := []exec.ActionType{
+		exec.ActionTypeUpdate,
+	}
+
+	t.Logf("\nPlan.Actions: %v", result.Actions)
+	t.Logf("\nPlan.Got: %v", result.Got)
+	t.Logf("\nPlan.Want: %v", result.Want)
+
+	err = expectActions(result.Actions, expectedActions)
+	if err != nil {
+		t.Fatalf("expectActions(_, _) = %v, want nil", err)
+	}
+	t.Log("\nstart NewSerialExecutor for update")
+	ex, err = exec.NewSerialExecutor(result.Actions, exec.TracerOption(&viz))
+	if err != nil {
+		t.Logf("exec.NewSerialExecutor err: %v", err)
+		return
+	}
+	res, err = ex.Run(context.Background(), theCloud)
+	if err != nil || res == nil {
+		t.Errorf("ex.Run(_,_) = ( %v, %v), want (*result, nil)", res, err)
+	}
+	t.Logf("exec.NewSerialExecutor finished, res: %v", res)
+	if len(res.Pending) > 0 {
+		t.Logf("Executor has %v, pending actions %v", len(res.Pending), res.Pending)
+		res, err = ex.Run(context.Background(), theCloud)
+		if err != nil || res == nil {
+			t.Errorf("ex.Run(_,_) = ( %v, %v), want (*result, nil)", res, err)
+		}
+	}
+	checkGCEResources(t, ctx, hcID, bsID, 15, 100)
+}
```

Comment:
these tests are extremely similar, can we refactor into table tests? it would make it easier to check all combinations:
1. HC update
2. BS update
3. HC+BS update
4. any resource we add in the future?

Reply:
done